plugins/connection/pct_remote.py

```diff
@@ -0,0 +1,396 @@
+#   remote_user: "'ansible'"
+#
+#
+# ----------------------
+# Playbook: playbook.yml
+# ----------------------
+---
+- hosts: lxc
+  tasks:
+    - debug:
+        msg: "This is coming from pct environment"
+"""
+
+
+import os
+import uuid
+from ansible.utils.display import Display
+from ansible.plugins.connection.paramiko_ssh import Connection as SSH_Connection
+from ansible.errors import AnsibleError
+from ansible import constants as C
+
+
+display = Display()
+
+
+def become_command():
+    """Helper function to get become_command """
+    return os.getenv('ANSIBLE_BECOME_METHOD', default=C.DEFAULT_BECOME_METHOD)
+
+
+class Connection(SSH_Connection):
+    ''' SSH based connections (paramiko) to Proxmox pct '''
+    transport = 'community.general.pct_remote'
+
+    def exec_command(self, cmd: str, in_data: bytes | None = None, sudoable: bool = True) -> tuple[int, bytes, bytes]:
+        ''' execute a command inside the proxmox container '''
+        cmd = ['/usr/sbin/pct', 'exec',
+               self.get_option('vmid'), '--', cmd]
+        if self.get_option('remote_user') != 'root':
+            cmd = [become_command()] + cmd
+        return super().exec_command(' '.join(cmd), in_data=in_data, sudoable=sudoable)
+
+    def put_file(self, in_path: str, out_path: str) -> None:
+        ''' transfer a file from local to remote '''
+        temp_dir = '/tmp/ansible_pct_' + str(uuid.uuid4()).replace('-', '')[:6]
+        temp_file_path = f'{temp_dir}/{os.path.basename(in_path)}'
+        try:
+            super().exec_command(f'mkdir -p {temp_dir}')
+            super().put_file(in_path, temp_file_path)
+            cmd = ['/usr/sbin/pct', 'push',
+                   self.get_option('vmid'), temp_file_path, out_path]
+            if self.get_option('remote_user') != 'root':
+                cmd = [become_command()] + cmd
+            super().exec_command(' '.join(cmd))
+        except Exception as e:
+            raise AnsibleError(
+                'failed to transfer file to %s!\n%s' % (out_path, e))
+        finally:
+            super().exec_command(f'rm -rf {temp_dir}')
+
+    def fetch_file(self, in_path: str, out_path: str) -> None:
+        ''' save a remote file to the specified path '''
+        temp_dir = '/tmp/ansible_pct_' + str(uuid.uuid4()).replace('-', '')[:6]
+        temp_file_path = f'{temp_dir}/{os.path.basename(in_path)}'
+        try:
+            super().exec_command(f'mkdir -p {temp_dir}')
+            cmd = ['/usr/sbin/pct', 'pull',
+                   self.get_option('vmid'), in_path, temp_file_path]
+            if self.get_option('remote_user') != 'root':
+                cmd = [become_command()] + cmd
+            super().exec_command(' '.join(cmd))
+            super().fetch_file(temp_file_path, out_path)
+        except Exception as e:
+            raise AnsibleError(
+                'failed to transfer file from %s!\n%s' % (in_path, e))
+        finally:
+            super().exec_command(f'rm -rf {temp_dir}')
```

Comment:
I would suggest replacing that logic with standard library's `tempfile.TemporaryDirectory()`:
```suggestion
        temp_dir = tempfile.TemporaryDirectory(prefix='ansible_pct_')
        temp_file_path = f'{temp_dir}/{os.path.basename(in_path)}'
        try:
            cmd = ['/usr/sbin/pct', 'pull', self.get_option('vmid'), in_path, temp_file_path]
            if self.get_option('remote_user') != 'root':
                cmd = [become_command()] + cmd
            super().exec_command(' '.join(cmd))
            super().fetch_file(temp_file_path, out_path)
        except Exception as e:
            raise AnsibleError(
                'failed to transfer file from %s!\n%s' % (in_path, e))
```
Same thing applies to the `put_file()` above.

See more at https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

Reply:
This is also what I thought at first, but when using `tempfile.TemporaryDirectory()` it will create a temp directory on the machine we are running the playbook from (`localhost`) and not the proxmox host (`remote`).

We could replace:

```
temp_dir = '/tmp/ansible_pct_' + str(uuid.uuid4()).replace('-', '')[:6]
```

with:

```
with tempfile.TemporaryDirectory() as tmp:
   temp_dir = '/tmp/ansible_pct_' + os.path.split(tmp)[1]
```

But I decided against it since we would momentarily create a temp directory on our `localhost`, which is unnecessary.

Reply:
Fair enough. But in that case, instead of calculating a temp dir name and creating it yourself, why not use `/usr/bin/mktemp` instead? It is part of coreutils (on Linuxmint, at least), it _should_ be available in any/most systems.

Reply:
shell plugins already handle this with mkdtemp function, which also centralizes all temp dirs/files on the remote and schedules a cleanup of the directory for when the task ends

Reply:
this is now obsolete